enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,821 @@
+```shell
+watch(DNSNameResolver)
+
+if match(event.type, "create") || match(event.type, "update"):
+
+	ensure_address_set(obj.spec.name)
+
+	ips = []
+	for each resolvedName in obj.status.resolvedNames:
+		add_ips(ips, resolvedName.ips)
+
+	set_ips_of_address_set(obj.spec.name, ips)
+
+else if match(event.type, "delete"):
+
+  	destroy_address_set(obj.spec.name)
+```
+
+* The OVN-K master(s) will watch for the events related to the `DNSNameResolver` CRs.
+* When the OVN-K master(s) receive(s) the create/update events, the master(s) will get the IP addresses related to the corresponding DNS name
+from the `.status` of the `DNSNameResolver` CR. The the OVN-K master(s) will then update the `AddressSet` for the DNS name, which is linked
+with the ACL rule(s) for the corresponding EgressFirewall rule(s).
+* When the OVN-K master(s) receive(s) the delete events, the OVN-K master(s) will delete the ACL rule(s) corresponding to the EgressFirewall
+rule(s) containing the DNS name(s).  The OVN-K master(s) will also delete the corresponding `AddressSet` for the corresponding DNS name.
+
+
+#### Variation [optional]
+
+
+### API Extensions
+
+The validation of [`DNSName`](https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/crd/egressfirewall/v1/types.go#L74-L76) field in
+`EgressFirewallDestination` will be updated to accept wildcard DNS names as well. It will be updated from `^([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.?$` which
+accepts only regular DNS names to `^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.?$`.
+
+````go
+// EgressFirewallDestination is the endpoint that traffic is either allowed or denied to
+type EgressFirewallDestination struct {
+	// ..
+
+	// dnsName is the domain name to allow/deny traffic to. If this is set, cidrSelector must be unset.
+	// For a wildcard DNS name, the '*' will match only one label. Additionally, only a single '*' can be
+	// used at the beginning of the wildcard DNS name. For example, '*.example.com' will match 'sub1.example.com'
+	// but won't match 'sub2.sub1.example.com'
+	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.?$
+	DNSName string `json:"dnsName,omitempty"`
+	// ..
+}
+````
+The details of the `DNSNameResolver` CRD can be found in the [Proposal](#proposal) section.
+
+
+### Implementation Details/Notes/Constraints [optional]
+
+The implementation changes needed for the proposed enhancement are documented in this section for each of the components.
+
+#### Cluster DNS Operator
+
+Cluster DNS Operator will deploy CoreDNS with the `ocp_dnsnameresolver` plugin enabled by adding it to the corefile. For the EgressFirewall rules to
+apply consistently, even for DNS names that are resolved by custom upstream nameservers, it will be added to all server blocks in the corefile. As the
+plugin will watch and update the `DNSNameResolver` CRs in the `network.openshift.io` api-group, proper RBAC permissions will be needed
+to be added to the `ClusterRole` for CoreDNS.
```

Comment:
Why not design a way for the EgressFirewall to configure this if it is activated?  If a customer doesn't even use the EgressFirewall, why would we need to make any changes in the plugin chain?

Reply:
Additionally, the feature will be behind a feature gate. If the feature gate is enabled, it means that the customer intends to use the feature and all the related components will be enabled including the new plugin.